### DISCUSSION TITLE
[Solved] Berry mqtt.susbcribe example

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#17752] [Q&A] [answered]
Hi, 
i m trying to use berry (low skill programming guy warning).
i dont understand how to get the payload i have subscribed to : 
should the payload be the argument of the function/closure (difficult to understand this concept despitre reading carefully the manual)

```
def recvmqtt()
print("function called")
end
mqtt.subscribe("cmnd/autotransfo/solarpower/#",recvmqtt())
```
the first time i run the code, i get the "function called" , i expect to get it periodically or at each change of the message value or something like that.
on the console i get this : 
`10:18:57.397 MQT: stat/autotransfo/RESULT = {"Command":"Unknown"}`

questions are : 
where is located the value that is sent in the mqtt message ? what argument should i give to recvmqtt to pass the value off the message ?
do i have tu run cyclically mqtt.subscribe ?  is this function ending or is it a loop ?
can someone past a small comprehensible example of this function (really simple, just passing the value to the  function)
i m also doubtful if it depends or not on the mqtt subscribe compile time option...

**Comment by sfromis** [comment, marked as answer]
Do read the [docs](https://tasmota.github.io/docs/Berry/#mqtt-module) about how to do.
> The function should return true if message was processed, false if not which will let the message flow to Tasmota eventually as a command.

Note that what you have in the subscribe is a function reference, not a call of the function, meaning:
`mqtt.subscribe("cmnd/autotransfo/solarpower",recvmqtt)`
(Also note that I removed that usually superfluous `/#` at the end, but that depends on your situation, of course)

You probably want to receive the data from the subscription instead of only knowing that "something" arrived.
`def recvmqtt(topic, idx, data, databytes)`

`idx` is not useful, and `databytes` is only really relevant for binary data.

You only need to subscribe once after boot (or if you unsubscribe, of course), then Tasmota will manage subscribing again if/when broker connection is established again after outages. You need no compile time options.

**Reply by sfromis** [reply]
More generally, it looks like what you want is to create a command. In that case **just forget** about using `mqtt.subscribe` at all, as Berry has a much better [feature](https://tasmota.github.io/docs/Berry/#functions-to-create-custom-tasmota-command) for that, `tasmota.add_cmd`

**Reply by ouinouin** [reply]
yes, y did in the mean time succeed with the command path, but this add a level of complexity to my low skills : manipulating object concepts.
however thanks for your reply since I managed to succeed, the missing point was the difference betwwen calling the function (function with the brackets) and doing reference to the function (function without brackets).
```
def mymqttdata(topic, idx, data, databytes)
  print("data: " , data)
end

import mqtt 
mqtt.subscribe("cmnd/autotransfo/solarpower",mymqttdata)
```
this snippet prints the payload carried by the topic cmnd/autotransfo/solarpower
@sfromis  thanks again for patience and always straight answer which let me progress in creating objects to  manage the energy in the house, today im experiencing a chinese autotransformer to change the power absorbed by my water boiler , coupled to a servomotor driven by tasmota dimmer, i then inject my solar panels instantaneous power in watts through mqtt , and finally close the feedback loop by reading the consumed power through a PZEM004 connected to the same tasmota :-) the servomotor adjusts the autotransformer do equalize produced solar energy and energy consumed by the boiler.

**Reply by sfromis** [reply]
If you want no error messages, do remember to follow it all:
> The function should return true if message was processed, false if not which will let the message flow to Tasmota eventually as a command.

And still, I'd really recommend registering it as a command, simpler to do than using MQTT, and the command also becomes available via console, serial, rules and so on.
